---
title: "Short Tandem Repeat"
output: 
  html_document:
    toc: FALSE 
    code_folding: hide

params:
  sample_id: ""
  input_bam: ""
  str_catalouge: ""
  reads_csv: ""
  filtered_reads_csv: ""
  clustering_summary_csv: ""
  em_summary_csv: ""
---

```{r Setup, include=FALSE}
# Setup
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  results = FALSE,
  out.width = "100%",
  fig.align = "center"
)

# Load libraries
library(tidyverse)
library(RColorBrewer)
library(knitr)
```

```{r Unpack input parameters}
# Sample information
sample_id <- params$sample_id
input_bam <- params$input_bam
str_catalouge <- params$str_catalouge

# Unpack input variables
reads_csv <- params$reads_csv
filtered_reads_csv <- params$filtered_reads_csv
clustering_summary_csv <- params$clustering_summary_csv
em_summary_csv <- params$em_summary_csv

print(params)

# TODO: Make variable names more descriptive
# Load input files
reads_df <- read_csv(reads_csv)
filtered_reads_df <- read_csv(filtered_reads_csv)
clustering_summary_df <- read_csv(clustering_summary_csv)
em_summary_df <- read_csv(em_summary_csv)
```

```{r Test exmaple data, eval=FALSE}
# Load data
reads_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.read_info.csv"
reads_df <- read_csv(reads_csv)

filtered_reads_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.filtered_reads_info.csv"
filtered_reads_df <- read_csv(filtered_reads_csv)

clustering_summary_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.haplotype_info.csv"
clustering_summary_df <- read_csv(clustering_summary_csv)

em_summary_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.summary.csv"
em_summary_df <- read_csv(em_summary_csv)
```

```{r Test exmaple data - local, eval=FALSE}
# Load data
reads_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/reads.csv"
reads_df <- read_csv(reads_csv)

filtered_reads_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/filtered_reads.csv"
filtered_reads_df <- read_csv(filtered_reads_csv)

clustering_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/haplotypes.csv"
clustering_summary_df <- read_csv(clustering_summary_csv)

em_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/summary.csv"
em_summary_df <- read_csv(em_summary_csv)
```


```{r Get destinct region}
# Get df of region information
regions <- reads_df %>%
  distinct(locus_id, locus_chrom, locus_start, locus_end, satellites_str, structure)
```


# Sample information
|||
|:----------------------------------|:--------------------------------------------|
| **Sample ID**                     | `r sample_id`                               |
| **Date**                          | `r format(Sys.time(), '%Y/%m/%d %H:%M %Z')` |
| **Analyzed BAM file**             | `r basename(input_bam)`                     |
| **STR catalouge**                 | `r basename(str_catalouge)`                 |

```{r Report, results = "asis"}
cat("\n\n")
cat("# Choose a region {.tabset .tabset-pills}")
cat("\n\n")

for (i in seq_len(nrow(regions))) {
  selected_region <- regions[i, ]

  # Filter data
  reads_region_df <- reads_df %>%
    semi_join(selected_region, by = join_by(locus_id)) %>%
    # Get number of kmers
    mutate(
      # Fix missing
      obs_kmer_string = if_else(is.na(obs_kmer_string), "", obs_kmer_string),
      n_kmer = str_count(obs_kmer_string, "-")
    ) %>%
    # Get max per em_haplotype
    group_by(em_haplotype) %>%
    mutate(n_kmer_haplotype = max(n_kmer)) %>%
    ungroup() %>%
    mutate(
      # Correct flanking reads
      start = case_when(
        alignment_type == "left_flanking" ~ -Inf,
        alignment_type == "right_flanking" ~ n_kmer_haplotype - n_kmer,
        alignment_type == "spanning" ~ -Inf
      ),
      end = case_when(
        alignment_type == "left_flanking" ~ n_kmer,
        alignment_type == "right_flanking" ~ Inf,
        alignment_type == "spanning" ~ Inf
      ),
      # Convert alignment_type to factor
      alignment_type = factor(alignment_type, levels = c("left_flanking", "spanning", "right_flanking"))
    ) %>%
    ungroup() %>%
    # Arrange reads by em_haplotype, alignment_type, mapping_start, length
    arrange(em_haplotype, alignment_type, start, n_kmer) %>%
    mutate(read_id = factor(row_number()))

  if (nrow(filtered_reads_df) > 0) {
    filtered_reads_region_df <- filtered_reads_df %>%
      semi_join(selected_region, by = join_by(locus_id))
  } else {
    filtered_reads_region_df <- data.frame()
  }

  em_summary_region_df <- em_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))

  # Create heading - Note this is a tab
  cat("\n\n")
  cat("##", selected_region$locus_id)
  cat("\n\n")


  cat("\n\n")
  cat("### Analysis summary \n")
  cat("\n\n")

  cat("\n\n")
  cat("#### STR summary")
  cat("\n\n")
  selected_region %>%
    mutate(
      goto_link = paste0("http://localhost:60151/goto?locus=", locus_chrom, ":", locus_start, "-", locus_end),
      igv_link = paste0("[Go to locus](", goto_link, ")"),
      structure = str_replace_all(structure, fixed("*"), fixed("\\*"))
    ) %>%
    select(
      "Locus ID" = locus_id,
      "Chrom" = locus_chrom,
      "Start" = locus_start,
      "End" = locus_end,
      "satellites" = satellites_str,
      "Structure" = structure,
      "IGV link" = igv_link
    ) %>%
    kable(
      align = "lcccccc"
    ) %>%
    print()

  cat("\n\n")
  cat("#### Read summary")
  cat("\n\n")

  data.frame(
    total_coverage = nrow(reads_region_df) + nrow(filtered_reads_region_df),
    filtered_reads = nrow(filtered_reads_region_df),
    analyzed_reads = nrow(reads_region_df),
    outliers = nrow(reads_region_df %>% filter(em_haplotype == "outlier")),
    haplotyped_reads = nrow(reads_region_df %>% filter(em_haplotype != "outlier"))
  ) %>%
    select(
      "Total coverage" = total_coverage,
      "Filtered" = filtered_reads,
      "Analyzed" = analyzed_reads,
      "Outliers" = outliers,
      "Used for haplotyping" = haplotyped_reads,
    ) %>%
    kable(
      align = "ccccc"
    ) %>%
    print()

  cat("\n\n")
  cat("#### Heterozygozity test summary \n")
  cat("\n\n")

  cat("This table shows the result from a statistical test for heterozygozity. The hypothesis ($H_A$) is that the sample is heterozygote i.e. that the data comes from two discrete normal distributions and the null hypothesis ($H_0$) is that the sample is homozygote i.e. that the data can be grouped into one gaussian cluster. The test performed is a log likelihood ratio test, where the statistic is assumed to be $\\chi^2$-distributed where the degrees of freedom is the difference in free paramteres under the two hypotheses. The value for the maximized log likelihood functions and the number of free parameteres under the two hypotheses, the test staistic, degrees of freedom, $p$-value and conclusion is shown in the following table.")

  em_summary_region_df %>%
    mutate(
      conclusion = if_else(
        is_significant,
        "Heterozygote",
        "Homozygote"
      )
    ) %>%
    select(
      "Log-likelihood Hetero" = log_lik_hetero,
      "Log-likelihood Homo" = log_lik_hom,
      "# Parameters Hetero" = n_par_hetero,
      "# Parameters Homo" = n_par_hom,
      "$\\chi^2$ Statistic" = statistic,
      "P-value" = p_value,
      "Significant" = is_significant,
      "Conclusion" = conclusion
    ) %>%
    kable(
      align = "cccccc"
    ) %>%
    print()

  cat("#### STR haplotype calls \n")

  res_region_df <- clustering_summary_df %>%
    semi_join(selected_region, by = join_by(locus_chrom, locus_start, locus_end)) %>%
    mutate_if(is.numeric, ~ round(., 2)) %>%
    select(
      "EM haplotype" = em_haplotype,
      "Count" = n,
      "satellite" = satellite,
      "Mean" = mean,
      "SD" = sd,
      "Median" = median,
      "IQR" = iqr,
    ) %>%
    kable(
      align = "lcccccc"
    ) %>%
    print()

  # Abacus plot
  cat("\n\n")
  cat("### Abacus plot {.tabset .tabset-pills}")
  cat("\n\n")

  cat("\n\n")
  cat("A visual representation illustrating the distribution of individual $k$-mers as colored beads along a linear sequence, with each sequence representing individual reads. The reads are grouped into haplotypes called by an unsupervised clustering algorithm (EM). Additionally, a separate group is designated for outlier reads. A read can be classified as an outlier either based on an unusual $k$-mer count or nucleotide sequence. The outlier detection is done using DBSCAN. A similar plot can be found under the *Phase* tab, where the grouping is based on the phasing of the alignment. A heatmap showing the concordance between the two groupings can be seen under *Concordance*.")
  cat("\n\n")

  # Wrangle STR data for abacus plot
  kmer_df <- reads_region_df %>%
    # Get one line per kmer
    mutate(
      kmer = str_split(obs_kmer_string, "-"),
      ref_kmer = str_split(ref_kmer_string, "-"),
      mod_kmer = str_split(mod_5mc_kmer_string, "-")
    ) %>%
    unnest_longer(
      col = c(kmer, ref_kmer, mod_kmer),
      values_to = "{col}",
      indices_to = "{col}_idx"
    ) %>%
    # Translated mod kmer to probability from ASCII value (33-43 -> 0,10,20,...,100)
    mutate(
      mod_prob = str_split(mod_kmer, "") %>% map(~ .x %>% map(~ .x %>%
        charToRaw() %>%
        as.numeric())) %>% map(~ max(unlist(.x) - 33)) %>% map(~ .x * 10) %>% unlist()
    )

  # Get top most common kmers
  generate_kmer_color_palette <- function(kmers) {
    # Make color palette for kmers
    col_pallete <- brewer.pal(max(3, min(11, length(kmers))), "Spectral")
    if (length(kmers) > 11) {
      kmer_colors <- colorRampPalette(colors = col_pallete)(length(kmers))
    } else {
      kmer_colors <- col_pallete[seq_along(kmers)]
    }
    names(kmer_colors) <- kmers
    return(kmer_colors)
  }

  # Function to create kmer plots
  create_kmer_plot <- function(kmer_df, color_by, n_colors) {
    # Fix data
    kmer_plot_df <- kmer_df %>%
      mutate(
        # Correct stort of right flanking reads
        cor_kmer_idx = if_else(
          alignment_type == "right_flanking",
          kmer_idx + start - 1,
          kmer_idx
        ),
        alignment_type_txt = if_else(
          alignment_type == "spanning", "Spanning", "Flanking"
        )
      ) %>%
      ungroup()

    # Start plotting
    plot <- ggplot()

    # Add horizontal lines for each read
    hz_lines_df <- kmer_plot_df %>%
      distinct(read_id, em_haplotype, alignment_type_txt, start, end, n_kmer_haplotype, n_kmer)

    plot <- plot +
      geom_segment(
        aes(
          x = start, xend = end,
          y = read_id, yend = read_id,
          alpha = alignment_type_txt
        ),
        lwd = 1,
        data = hz_lines_df
      ) +
      scale_alpha_manual(values = c("Spanning" = 1, "Flanking" = 0.4)) +
      labs(alpha = "Alignment type")

    # Add points for each kmer
    if (color_by == "kmer") {
      # Get top kmers
      top_kmers <-
        kmer_plot_df %>%
        count(kmer) %>%
        # Remove unique kmers
        filter(n > 1) %>%
        # Get top kmers
        arrange(desc(n)) %>%
        head(n_colors) %>%
        pull(kmer)

      # Get color palette
      color_palette <- generate_kmer_color_palette(top_kmers)

      # Add points and color by kmer
      plot <- plot +
        geom_point(
          aes(
            x = cor_kmer_idx, y = read_id,
            col = factor(kmer, levels = top_kmers)
          ),
          size = 5,
          data = kmer_plot_df
        ) +
        scale_color_manual(values = color_palette) +
        labs(color = "K-mer\n(ord. by freq.)")
    } else if (color_by == "mod") {
      # Add points and color by mod probability
      plot <- plot +
        geom_point(
          aes(x = cor_kmer_idx, y = read_id, col = mod_prob),
          size = 5,
          data = kmer_plot_df
        ) +
        scale_color_gradient(low = "red", high = "blue", limits = c(0, 100)) +
        labs(color = "5mC (prob.)")
    } else {
      # Error
      stop("Invalid color_by argument.")
    }

    # Add details to plot
    plot <- plot +
      facet_grid(
        rows = vars(em_haplotype),
        scales = "free_y",
        space = "free_y"
      ) +
      labs(
        title = "K-mer overview",
        x = "K-mer index",
        y = "Read ID",
      ) +
      scale_y_discrete(limits = rev)
    return(plot)
  }

  cat("\n\n")
  cat("#### Sequence")
  cat("\n\n")
  kmer_plot_df <- kmer_df %>%
    filter(em_haplotype != "outlier")
  kmer_plot <- create_kmer_plot(kmer_plot_df, "kmer", 15)
  print(kmer_plot)


  cat("\n\n")
  cat("#### Reference")
  cat("\n\n")
  kmer_plot_df <- kmer_df %>%
    filter(em_haplotype != "outlier") %>%
    mutate(
      kmer = ref_kmer
    )
  kmer_plot <- create_kmer_plot(kmer_plot_df, "kmer", 15)
  print(kmer_plot)


  cat("\n\n")
  cat("#### Methylation (5mC)")
  cat("\n\n")
  kmer_plot_df <- kmer_df %>%
    filter(em_haplotype != "outlier")
  kmer_plot <- create_kmer_plot(kmer_plot_df, "mod", 11)
  print(kmer_plot)


  cat("\n\n")
  cat("#### Outliers")
  cat("\n\n")
  kmer_plot_df <- kmer_df %>%
    filter(em_haplotype == "outlier")

  if (nrow(kmer_plot_df) > 0) {
    kmer_plot <- create_kmer_plot(kmer_plot_df, "kmer", 15)
    print(kmer_plot)
  } else {
    cat("There are no outliers for this locus.")
  }

  cat("\n\n")
  cat("### Additional read information")
  cat("\n\n")

  cat("A table with additional information for the reads visualized above.")

  reads_region_df %>%
    select(
      "Read ID" = read_id,
      "Query Name" = query_name,
      "Strand" = strand,
      "Alignment Type" = alignment_type,
      "EM Haplotype" = em_haplotype,
      "Outlier Reason" = outlier_reason,
      "Median STR Qual" = median_str_qual,
      "Kmer Count" = kmer_count_str
    ) %>%
    mutate_if(is.numeric, ~ round(., 3)) %>%
    kable(
      align = "lcccccc"
    ) %>%
    print()

  cat("\n\n")
  cat("### Filtered reads")
  cat("\n\n")

  cat("A table with information about the filtered reads and the reason why they were removed.\n")

  if (nrow(filtered_reads_region_df) > 0) {
    filtered_reads_region_df %>%
      select(
        "Query Name" = query_name,
        "Strand" = strand,
        "Error flags" = error_flags
      ) %>%
      mutate_if(is.numeric, ~ round(., 3)) %>%
      kable(
        align = "lccc"
      ) %>%
      print()
  } else {
    cat("There is no filtered reads for this locus. \n")
  }

  cat("\n\n")
  cat("### Histogram plots {.tabset .tabset-pills}")
  cat("\n\n")

  cat("A histogram showing the number of occurences of the individual $k$-mers. Is more than one $k$-mer is present in the pattern one histogram is shown for each $k$-mer. Reads that are removed as outliers are not shown in the plot.")

  res_region_df <- clustering_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))

  # Plot count
  generate_length_plot <- function(counts_df, density_df, satellite, BIN_SZ) {
    # Bin count to min max 100 bins
    NBINS <- min(100, max(counts_df$count))
    BIN_SZ <- ceiling(max(counts_df$count) / NBINS)

    counts_df <- counts_df %>%
      mutate(
        count = count %/% BIN_SZ * BIN_SZ + BIN_SZ / 2
      ) %>%
      group_by(alignment_type_txt, count) %>%
      summarise(n = sum(n), .groups = "drop")

    length_plot <- ggplot(counts_df) +
      # Add histogram of STR length
      geom_col(
        aes(
          x = count,
          y = n,
          fill = alignment_type_txt
        ),
        width = BIN_SZ
      ) +
      labs(
        title = paste0("Satellite: ", satellite),
        x = "STR length",
        y = "Count"
      ) +
      scale_x_continuous(breaks = function(x) unique(round(sort(pretty(x))))) +
      scale_y_continuous(breaks = function(x) unique(round(sort(pretty(x))))) +
      scale_fill_manual(values = c("Spanning" = "grey30", "Flanking" = "grey70"))

    # Add distributions
    if (nrow(density_df) > 0) {
      # Fix colors
      em_haplotype_colors <- brewer.pal(3, "Set1")
      names(em_haplotype_colors) <- c("hom", "h1", "h2")

      # Scale density to max count
      if (nrow(density_df) > 0) {
        density_df <- density_df %>%
          group_by(em_haplotype) %>%
          mutate(
            d = d / max(d) * max(counts_df$n)
          ) %>%
          ungroup()
      }

      length_plot <- length_plot +
        geom_line(
          data = density_df,
          aes(x = x, y = d, col = em_haplotype)
        ) +
        scale_color_manual(values = em_haplotype_colors)
    }

    return(length_plot)
  }

  # Get parameters from EM summary
  em_het_par <- em_summary_region_df %>%
    mutate(
      het_pi_h1 = as.character(pi),
      het_pi_h2 = as.character(1 - pi)
    ) %>%
    select(het_pi_h1, het_pi_h2, het_mean_h1, het_mean_h2, het_sd_h1, het_sd_h2) %>%
    # Pivot mean and var into long format. Use h1 and h2 as suffix
    pivot_longer(
      cols = everything(),
      names_pattern = "het_(pi|mean|sd)_(h[0-9])",
      names_to = c("stat", "em_haplotype")
    ) %>%
    # Split value for each kmer
    mutate(
      value = str_split(value, ";"),
    ) %>%
    # Uneast value
    unnest_longer(value, indices_to = "satellite_idx") %>%
    # Convert to numeric
    mutate(
      value = as.numeric(value)
    ) %>%
    # Pivot stat into columns
    pivot_wider(
      names_from = stat,
      values_from = value
    )

  em_hom_par <- em_summary_region_df %>%
    select(hom_mean, hom_sd) %>%
    pivot_longer(
      cols = everything(),
      names_pattern = "hom_(mean|sd)",
      names_to = "stat",
      values_to = "value"
    ) %>%
    # Split value for each kmer
    mutate(
      value = str_split(value, ";"),
    ) %>%
    # Uneast value
    unnest_longer(value, indices_to = "satellite_idx") %>%
    # Convert to numeric
    mutate(
      value = as.numeric(value)
    ) %>%
    # Pivot stat into columns
    pivot_wider(
      names_from = stat,
      values_from = value
    ) %>%
    # Add em_haplotype
    mutate(
      em_haplotype = "hom",
      pi = 1
    )


  density_df <-
    bind_rows(em_het_par, em_hom_par) %>%
    rowwise() %>%
    mutate(
      plot_sd = pmax(sd, 0.2),
      x = list(seq(mean - 5 * plot_sd, mean + 5 * plot_sd, length.out = 101)),
    ) %>%
    unnest(x) %>%
    group_by(em_haplotype) %>%
    mutate(
      d = dnorm(x, mean = mean, sd = plot_sd)
    )

  length_df <- reads_region_df %>%
    mutate(
      count = str_split(kmer_count_str, "-")
    ) %>%
    unnest_longer(col = count, indices_to = "satellite_idx") %>%
    mutate(
      count = as.integer(count),
      alignment_type_txt = if_else(
        alignment_type == "spanning", "Spanning", "Flanking"
      )
    ) %>%
    select(
      em_haplotype, alignment_type_txt, satellite_idx, count
    )

  satellites <- str_split(selected_region$satellites_str, "-", simplify = TRUE)
  for (j in seq_along(satellites)) {
    # Get satellite
    satellite <- satellites[j]

    # Filter data

    counts_index_df <- length_df %>%
      filter(
        satellite_idx == !!j
      ) %>%
      count(
        em_haplotype, alignment_type_txt, count
      )

    density_index_df <-
      density_df %>%
      filter(satellite_idx == !!j)

    cat("\n\n")
    cat("#### Grouped")
    cat("\n\n")
    length_plot_df <- counts_index_df %>%
      filter(em_haplotype != "outlier")
    if (nrow(length_plot_df) > 0) {
      length_plot <- generate_length_plot(length_plot_df, density_index_df, satellite, BIN_SZ)
      print(length_plot)
    } else {
      cat("There are no grouped reads for this locus.")
    }

    cat("\n\n")
    cat("#### Outliers")
    cat("\n\n")
    length_plot_df <- counts_index_df %>%
      filter(em_haplotype == "outlier")

    if (nrow(length_plot_df) > 0) {
      length_plot <- generate_length_plot(length_plot_df, data.frame(), satellite, BIN_SZ)
      print(length_plot)
    } else {
      cat("There are no outliers for this locus.")
    }
  }
}
```
