---
title: "Short Tandem Repeat (STR) analysis"
output: 
  html_document:
    toc: FALSE 
    code_folding: hide

params:
  sample_id: ""
  input_bam: ""
  str_catalouge: ""
  reads_csv: ""
  consensus_csv: ""
  filtered_reads_csv: ""
  clustering_summary_csv: ""
  test_summary_csv: ""
  par_summary_csv: ""
---

```{r Setup, include=FALSE}
# Setup
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  results = FALSE,
  out.width = "100%",
  fig.align = "center"
)

options(knitr.table.format = "html")

# Load libraries
library(tidyverse)
library(RColorBrewer)
library(knitr)
library(kableExtra)
```

```{r Unpack input parameters}
# Sample information
sample_id <- params$sample_id
input_bam <- params$input_bam
str_catalouge <- params$str_catalouge

# Unpack input variables
reads_csv <- params$reads_csv
filtered_reads_csv <- params$filtered_reads_csv
consensus_csv <- params$consensus_csv
clustering_summary_csv <- params$clustering_summary_csv
test_summary_csv <- params$test_summary_csv
par_summary_csv <- params$par_summary_csv
```

```{r Test exmaple data, eval=FALSE}
reads_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.read_info.csv"
filtered_reads_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.filtered_reads_info.csv"
clustering_summary_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.haplotype_info.csv"
test_summary_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.summary.csv"
```

```{r Test exmaple data - local, eval=FALSE}
# Variables
sample_id <- "test"
input_bam <- "path/to/bam"
str_catalouge <- "path/to/str_catalogue"

# Data files
reads_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/reads.csv"
filtered_reads_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/filtered_reads.csv"
consensus_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/consensus.csv"
clustering_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/haplotypes.csv"
test_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/summary.csv"
par_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/par_summary.csv"


reads_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/coriell_samples/coriell_sample_outputs/abacus_output/reads.csv"
filtered_reads_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/coriell_samples/coriell_sample_outputs/abacus_output/filtered_reads.csv"
consensus_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/coriell_samples/coriell_sample_outputs/abacus_output/consensus.csv"
clustering_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/coriell_samples/coriell_sample_outputs/abacus_output/haplotypes.csv"
test_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/coriell_samples/coriell_sample_outputs/abacus_output/summary.csv"
par_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/coriell_samples/coriell_sample_outputs/abacus_output/par_summary.csv"
```

```{r Load data}
reads_df <- read_csv(reads_csv)
filtered_reads_df <- read_csv(filtered_reads_csv)
consensus_df <- read_csv(consensus_csv)
clustering_summary_df <- read_csv(clustering_summary_csv)
test_summary_df <- read_csv(test_summary_csv)
all_par_summary_df <- read_csv(par_summary_csv)
```

```{r Constants}
ALIGNMENT_TYPES <- c("left_flanking", "spanning", "right_flanking")

# Colors
HEADER_BACKGROUND <- "grey87"
STRIPED_BACKGROUND <- "grey97"
```

```{r Functions}
wrangle_reads_data <- function(reads_df) {
  reads_df %>%
    # Get number of kmers
    mutate(
      # Fix missing
      obs_kmer_string = if_else(is.na(obs_kmer_string), "", obs_kmer_string),
      # Count kmers
      n_kmer = str_count(obs_kmer_string, "-") + 1,
      # Convert alignment_type to factor
      alignment_type_fct = factor(alignment_type, levels = ALIGNMENT_TYPES),
      alignment_type_txt = if_else(
        alignment_type == "spanning", "Spanning", "Flanking"
      ),
      alignment_subtype_txt = case_when(
        alignment_type == "left_flanking" ~ "Left flanking",
        alignment_type == "right_flanking" ~ "Right flanking",
        alignment_type == "spanning" ~ "Spanning"
      )
    ) %>%
    # Arrange reads by haplotype, alignment_type, mapping_start, length and sequence
    arrange(
      haplotype,
      alignment_type_fct,
      if_else(
        alignment_type == "right_flanking",
        desc(n_kmer),
        n_kmer
      ),
      obs_kmer_string
    ) %>%
    ungroup() %>%
    # Assign read_id
    mutate(read_id = row_number())
}

# Wrangle STR data for abacus plot
mod_ascii_to_prob <- function(mod_ascii) {
  # Check if mod_ascii is empty
  if (nchar(mod_ascii) == 0) {
    return(0)
  }
  # Convert to numeric
  mod_prob <- mod_ascii %>%
    charToRaw() %>%
    as.numeric()
  # Correct for ASCII values
  mod_prob <- (mod_prob - 33) * 10
  # Remove values outside 0-100
  mod_prob <- mod_prob[mod_prob >= 0 & mod_prob <= 100]
  # Return NA if no values
  if (length(mod_prob) == 0) {
    return(NA)
  }
  # Set mod_prob to max value
  mod_prob <- max(mod_prob)
  return(mod_prob)
}

wrangle_reads_to_kmers <- function(reads_df) {
  reads_df %>%
    # Get one line per kmer
    mutate(
      kmer = str_split(obs_kmer_string, "-"),
      ref_kmer = str_split(ref_kmer_string, "-"),
      mod_kmer = str_split(mod_5mc_kmer_string, "-")
    ) %>%
    unnest_longer(
      col = c(kmer, ref_kmer, mod_kmer),
      values_to = "{col}",
      indices_to = "{col}_idx"
    ) %>%
    # Fix NA values
    mutate(
      kmer = if_else(is.na(kmer), "", kmer),
      ref_kmer = if_else(is.na(ref_kmer), "", ref_kmer),
      mod_kmer = if_else(is.na(mod_kmer), "", mod_kmer)
    ) %>%
    # Translated mod kmer to probability from ASCII value (33-43 -> 0,10,20,...,100)
    rowwise() %>%
    mutate(
      mod_prob = mod_ascii_to_prob(mod_kmer)
    )
}

# Make color palette for kmers
generate_kmer_color_palette <- function(kmers) {
  # Set max number of colors
  max_cols <- 12

  # Trim kmers if too many
  if (length(kmers) > max_cols) {
    kmers <- kmers[1:max_cols]
  }

  # Get color palette
  full_paired_pallete <- brewer.pal(max_cols, "Paired")
  reordered_pallete <- c(full_paired_pallete[(1:6) * 2], full_paired_pallete[(1:6) * 2 - 1])

  kmer_colors <- reordered_pallete[seq_along(kmers)]
  names(kmer_colors) <- kmers

  return(kmer_colors)
}

# Function to create kmer plots
create_kmer_plot <- function(kmer_df, color_by, kmer_color_palette = NULL) {
  kmer_plot_df <- kmer_df %>%
    # Get max per haplotype
    group_by(haplotype) %>%
    mutate(max_kmer_haplotype = max(n_kmer)) %>%
    ungroup() %>%
    # Get start and end for flanking reads
    mutate(
      start = case_when(
        alignment_type == "left_flanking" ~ -Inf,
        alignment_type == "right_flanking" ~ max_kmer_haplotype - n_kmer + 1,
        alignment_type == "spanning" ~ -Inf
      ),
      end = case_when(
        alignment_type == "left_flanking" ~ n_kmer,
        alignment_type == "right_flanking" ~ Inf,
        alignment_type == "spanning" ~ Inf
      ),
    ) %>%
    # Correct start and end of flanking reads
    mutate(
      cor_kmer_idx = if_else(
        alignment_type == "right_flanking",
        kmer_idx + start - 1,
        kmer_idx
      )
    ) %>%
    ungroup()

  # Start plotting
  plot <- ggplot()

  # Add horizontal lines for each read
  hz_lines_df <- kmer_plot_df %>%
    distinct(read_id, alignment_type_txt, start, end, haplotype)

  plot <- plot +
    geom_segment(
      aes(
        x = start, xend = end,
        y = read_id, yend = read_id,
        alpha = alignment_type_txt
      ),
      lwd = 1,
      data = hz_lines_df
    ) +
    scale_alpha_manual(values = c("Spanning" = 1, "Flanking" = 0.4)) +
    labs(alpha = "Alignment type")

  # Get "dimensionality" of plot
  n_reads <- nrow(hz_lines_df)
  n_kmers <- max(kmer_plot_df$cor_kmer_idx)

  # Add points for each kmer
  if (color_by == "kmer" || color_by == "ref_kmer") {
    # If ref_kmer is selected, use ref_kmer instead of kmer
    if (color_by == "ref_kmer") {
      kmer_plot_df <- kmer_plot_df %>%
        mutate(kmer = ref_kmer)
    }

    # Add "Other" (grey) to color palette in end
    kmer_color_palette <- c(kmer_color_palette, c("Other" = "grey"))

    kmer_plot_df <- kmer_plot_df %>%
      mutate(
        # Change non-top kmers to "Other"
        kmer = if_else(kmer %in% names(kmer_color_palette), kmer, "Other"),
        # Convert to factor
        kmer = factor(kmer, levels = names(kmer_color_palette))
      )

    # Add points/tiles and color by kmer
    if (n_reads < 50 && n_kmers < 60) {
      plot <- plot +
        geom_point(
          aes(
            x = cor_kmer_idx, y = read_id,
            col = kmer
          ),
          size = 5,
          data = kmer_plot_df
        ) +
        # Add color palette
        scale_color_manual(values = kmer_color_palette) +
        labs(color = "K-mer\n(ord. by freq.)")
    } else {
      plot <- plot +
        geom_tile(
          aes(
            x = cor_kmer_idx, y = read_id,
            fill = kmer
          ),
          data = kmer_plot_df
        ) +
        # Add color palette
        scale_fill_manual(values = kmer_color_palette) +
        labs(fill = "K-mer\n(ord. by freq.)")
    }
  } else if (color_by == "mod") {
    # Add points/tiles and color by mod probability
    if (n_reads < 50 && n_kmers < 60) {
      plot <- plot +
        geom_point(
          aes(x = cor_kmer_idx, y = read_id, col = mod_prob),
          size = 5,
          data = kmer_plot_df
        ) +
        # Add color gradient with low (0) = "blue", high (100) = "red", NA = "grey" (similar to IGV)
        scale_color_gradient(
          low = "blue",
          high = "red",
          na.value = "grey",
          limits = c(0, 100)
        ) +
        labs(color = "5mC (prob.)")
    } else {
      plot <- plot +
        geom_tile(
          aes(x = cor_kmer_idx, y = read_id, fill = mod_prob),
          data = kmer_plot_df
        ) +
        # Add color gradient with low (0) = "blue", high (100) = "red", NA = "grey" (similar to IGV)
        scale_fill_gradient(
          low = "blue",
          high = "red",
          na.value = "grey",
          limits = c(0, 100)
        ) +
        labs(fill = "5mC (prob.)")
    }
  } else {
    # Error
    stop("Invalid color_by argument.")
  }

  # Add details to plot
  min_kmer_idx <- min(kmer_plot_df$cor_kmer_idx)
  vertical_fixer_df <- kmer_plot_df %>%
    distinct(read_id, haplotype) %>%
    group_by(haplotype) %>%
    summarise(
      min_read_idx = min(read_id),
      max_read_idx = max(read_id),
      # Group size
      group_size = n(),
    ) %>%
    ungroup() %>%
    mutate(
      # Add space between groups
      min_read_idx = if_else(group_size < 10, min_read_idx - 0.5, min_read_idx),
      max_read_idx = if_else(group_size < 10, max_read_idx + 0.5, max_read_idx),
      x = 0.5
    )

  plot <- plot +
    facet_grid(
      rows = vars(haplotype),
      scales = "free_y",
      space = "free_y"
    ) +
    # Fix y axis with "fake" data, i.e. transparent lines
    geom_segment(
      aes(
        x = x, xend = x,
        y = min_read_idx, yend = max_read_idx,
      ),
      data = vertical_fixer_df,
      alpha = 0,
    ) +
    labs(
      x = "K-mer index",
      y = "Read ID",
    ) +
    # Breaks for y: Every integer, pretty, start at 1
    scale_y_reverse(breaks = function(x) unique(round(sort(pmax(min_kmer_idx, pretty(x)))))) +
    # Remove "minor" ticks
    theme(
      panel.grid.minor = element_blank()
    )

  return(plot)
}

add_colors_to_consensus_strings <- function(consensus_strings, kmer_color_palette) {
  consensus_strings_parts <- str_split(consensus_strings, "-")
  colored_consensus_strings <- c()
  for (part in consensus_strings_parts) {
    # Extract counts
    counts <- str_extract(part, "[0-9]+")

    # Extract kmers
    kmers <- str_extract(part, "[A-Z]+")

    # Add color
    colored_consensus_string <- ""
    for (i in seq_along(counts)) {
      count <- counts[i]
      kmer <- kmers[i]

      # Add color and make bold
      if (kmer %in% names(kmer_color_palette)) {
        colored_kmer <- paste0("<span style='color:", kmer_color_palette[kmer], "; font-weight: bold;'>", kmer, "</span>")
      } else {
        colored_kmer <- kmer
      }

      # Add to colored string
      colored_consensus_string <- paste0(colored_consensus_string, "-", count, "(", colored_kmer, ")")
    }

    # Remove first "-"
    colored_consensus_string <- substr(colored_consensus_string, 2, nchar(colored_consensus_string))

    # Add to list
    colored_consensus_strings <- c(colored_consensus_strings, colored_consensus_string)
  }
  return(colored_consensus_strings)
}

generate_length_plot <- function(counts_df, density_df, satellite, BIN_SZ) {
  # Bin count to min max 100 bins
  NBINS <- min(100, max(counts_df$count))
  BIN_SZ <- ceiling(max(counts_df$count) / NBINS)

  counts_df <- counts_df %>%
    mutate(
      count = count %/% BIN_SZ * BIN_SZ
    ) %>%
    group_by(alignment_type_txt, count) %>%
    summarise(n = sum(n), .groups = "drop")

  length_plot <- ggplot(counts_df) +
    # Add histogram of STR length
    geom_col(
      aes(
        x = count,
        y = n,
        fill = alignment_type_txt
      ),
      width = BIN_SZ
    ) +
    scale_x_continuous(breaks = function(x) unique(round(sort(pretty(x))))) +
    scale_y_continuous(breaks = function(x) unique(round(sort(pretty(x))))) +
    scale_fill_manual(values = c("Spanning" = "grey30", "Flanking" = "grey70"))

  # Add distributions
  if (nrow(density_df) > 0) {
    # Fix colors
    haplotype_colors <- brewer.pal(3, "Set1")
    names(haplotype_colors) <- c("hom", "h1", "h2")
    # Fix line type
    haplotype_linetype <- c("hom" = "dashed", "h1" = "solid", "h2" = "solid")

    # Scale density to max count
    if (nrow(density_df) > 0) {
      density_df <- density_df %>%
        group_by(haplotype) %>%
        mutate(
          d = d / max(d) * max(counts_df$n)
        ) %>%
        ungroup()
    }

    length_plot <- length_plot +
      geom_line(
        data = density_df,
        aes(x = x, y = d, col = haplotype, linetype = haplotype),
        linewidth = 0.5
      ) +
      scale_color_manual(values = haplotype_colors) +
      scale_linetype_manual(values = haplotype_linetype, guide = "none")
  }

  # Add title, labels and legend
  length_plot <- length_plot +
    # Add title and labels
    labs(
      title = paste("Satellite:", satellite),
      x = "STR length",
      y = "Count"
    ) +
    # Name the legend
    guides(
      fill = guide_legend(title = "Alignment type"),
      col = guide_legend(title = "Haplotype")
    )

  return(length_plot)
}
```

```{r Sample information table, results = TRUE}
# Create table
data.frame(
  sample_id = sample_id,
  date = format(Sys.time(), "%Y/%m/%d %H:%M %Z"),
  bam = basename(input_bam),
  json = basename(str_catalouge)
) %>%
  select(
    "Sample ID" = sample_id,
    "Date" = date,
    "BAM file" = bam,
    "STR catalouge" = json
  ) %>%
  # Transpose
  t() %>%
  kbl(align = "l") %>%
  kable_styling("bordered", full_width = TRUE) %>%
  # Make first column grey
  column_spec(1, background = HEADER_BACKGROUND, bold = TRUE)
```

# Test summary

```{r Summary Table, results = TRUE}
test_df <- test_summary_df %>%
  select(locus_id, is_significant)

signif_par_summary_df <- all_par_summary_df %>%
  # Filter on significant estimates
  left_join(test_df, by = join_by(locus_id)) %>%
  filter((is_significant & haplotype != "hom") | (!is_significant) & haplotype == "hom") %>%
  # Arrange by satellite and haplotype
  arrange(locus_id, idx, haplotype)

background_row_index <- signif_par_summary_df %>%
  # Get row index
  group_by(locus_id) %>%
  summarise(
    n = n(),
  ) %>%
  mutate(
    a = cumsum(n)
  ) %>%
  # Remove every other group
  filter(row_number() %% 2 == 1) %>%
  rowwise() %>%
  mutate(
    x = list((a - n + 1):a)
  ) %>%
  unnest(x) %>%
  pull(x)

signif_par_summary_df %>%
  # Round to 2 decimals
  mutate_if(is.numeric, ~ round(., 2)) %>%
  # Select parameters, round and print
  select(
    "Locus ID" = locus_id,
    "Satellite" = satellite,
    "Haplotype" = haplotype,
    "Count estimate" = mean,
    "95% CI Lower" = mean_lower,
    "95% CI Upper" = mean_upper
  ) %>%
  kbl(
    align = "c"
  ) %>%
  column_spec(1, italic = TRUE) %>%
  kable_styling("bordered", full_width = TRUE) %>%
  collapse_rows(columns = c(1, 2), valign = "middle", target = 1, ) %>%
  row_spec(
    0,
    background = HEADER_BACKGROUND
  ) %>%
  # Add background color to every other row
  row_spec(
    background_row_index,
    background = STRIPED_BACKGROUND
  )
```

```{r Build report, results = "asis"}
cat("\n\n")
cat("# Choose a region {.tabset .tabset-pills}")
cat("\n\n")

# Get destinct regions (for looping)
regions <- reads_df %>%
  distinct(locus_id, locus_chrom, locus_start, locus_end, satellites_str, structure) %>%
  arrange(locus_id)

for (i in seq_len(nrow(regions))) {
  selected_region <- regions[i, ]

  # Filter data to selected region
  reads_region_df <- reads_df %>%
    semi_join(selected_region, by = join_by(locus_id)) %>%
    wrangle_reads_data()

  if (nrow(filtered_reads_df) > 0) {
    filtered_reads_region_df <- filtered_reads_df %>%
      semi_join(selected_region, by = join_by(locus_id))
  } else {
    filtered_reads_region_df <- data.frame()
  }

  test_summary_region_df <- test_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))

  par_summary_region_df <- all_par_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))

  signif_par_summary_region_df <- signif_par_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))

  clustering_summary_region_df <- clustering_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))

  # Calculate coverage
  n_haplotyped_reads <- nrow(reads_region_df %>% filter(haplotype != "outlier"))
  n_outlier_reads <- nrow(reads_region_df %>% filter(haplotype == "outlier"))
  n_filtered_reads <- nrow(filtered_reads_region_df)


  # Create heading - Note this is a tab
  cat("\n\n")
  cat("##", selected_region$locus_id)
  cat("\n\n")

  selected_region %>%
    mutate(
      goto_link = paste0("http://localhost:60151/goto?locus=", locus_chrom, ":", locus_start, "-", locus_end),
      igv_link = paste0("[Go to locus](", goto_link, ")"),
      structure = str_replace_all(structure, fixed("*"), fixed("\\*"))
    ) %>%
    select(
      "Locus ID" = locus_id,
      "Chrom" = locus_chrom,
      "Start" = locus_start,
      "End" = locus_end,
      "Structure" = structure,
      "IGV link" = igv_link
    ) %>%
    kbl(
      align = "c"
    ) %>%
    kable_styling("bordered", full_width = TRUE) %>%
    row_spec(
      0,
      background = HEADER_BACKGROUND
    ) %>%
    cat()

  # Abacus plot
  cat("\n\n")
  cat("### Abacus plot {.tabset .tabset-pills}")
  cat("\n\n")

  cat("\n\n")
  cat("A visual representation illustrating the distribution of individual $k$-mers as colored *beads* along a linear sequence, with each horizontal sequence representing individual reads. The reads are grouped into haplotypes called by an unsupervised clustering algorithm (EM). Additionally, a separate group is designated for outlier reads. A read can be classified as an outlier either based on an unusual $k$-mer count or nucleotide sequence. There are 3 versions of the plot where the beads are colored by the observed $k$-mer, the reference $k$-mer, and the methylation probability of the $k$-mer. Spanning and flanking reads are shown as black and grey lines, respectively.")
  cat("\n\n")

  kmer_df <- wrangle_reads_to_kmers(reads_region_df)

  # Split data
  kmer_haplotyped_df <- kmer_df %>%
    filter(haplotype != "outlier")

  kmer_outlier_df <- kmer_df %>%
    filter(haplotype == "outlier")

  no_haplotype_reads <- nrow(kmer_haplotyped_df) == 0
  no_outlier_reads <- nrow(kmer_outlier_df) == 0

  # Get color palette
  # Get kmers that should be prioritized for color
  # Ref kmers
  ref_kmers <- kmer_haplotyped_df %>%
    filter(!is.na(ref_kmer)) %>%
    count(ref_kmer) %>%
    arrange(desc(n)) %>%
    pull(ref_kmer)

  # Most observed kmers
  top_obs_kmers <-
    kmer_haplotyped_df %>%
    # Remove outlier reads, NA and empty kmers
    filter(haplotype != "outlier", !is.na(kmer), kmer != "") %>%
    count(kmer) %>%
    # Remove unique kmers (assumed to be errors)
    filter(n > 1) %>%
    # Get top kmers
    arrange(desc(n)) %>%
    pull(kmer) %>%
    # Remove kmers that are in ref_kmers
    setdiff(ref_kmers)

  # Select kmers for color palette
  n_colors <- 12
  top_kmers <- c(ref_kmers, top_obs_kmers) %>%
    unique() %>%
    head(n_colors)

  kmer_color_palette <- generate_kmer_color_palette(top_kmers)

  cat("\n\n")
  cat("#### Sequence (N=", n_haplotyped_reads, ")", sep = "")
  cat("\n\n")

  if (no_haplotype_reads) {
    cat("There are no haplotyped reads for this locus.")
  } else {
    kmer_plot <- create_kmer_plot(kmer_haplotyped_df, "kmer", kmer_color_palette)
    print(kmer_plot)
  }


  cat("\n\n")
  cat("#### Reference")
  cat("\n\n")

  if (no_haplotype_reads) {
    cat("There are no haplotyped reads for this locus.")
  } else {
    kmer_plot <- create_kmer_plot(kmer_haplotyped_df, "ref_kmer", kmer_color_palette)
    print(kmer_plot)
  }

  cat("\n\n")
  cat("#### Methylation (5mC)")
  cat("\n\n")

  if (no_haplotype_reads) {
    cat("There are no haplotyped reads for this locus.")
  } else if (all(is.na(kmer_haplotyped_df$mod_prob))) {
    cat("There are no methylation information for this locus.")
  } else {
    kmer_plot <- create_kmer_plot(kmer_haplotyped_df, "mod", 11)
    print(kmer_plot)
  }

  cat("\n\n")
  cat("#### Outliers (N=", n_outlier_reads, ")", sep = "")
  cat("\n\n")

  if (no_outlier_reads) {
    cat("There are no outliers for this locus.")
  } else {
    # Outlier color pallette
    top_outlier_kmers <- kmer_outlier_df %>%
      filter(!is.na(kmer)) %>%
      count(kmer) %>%
      arrange(desc(n)) %>%
      pull(kmer) %>%
      head(n_colors)

    outlier_color_palette <- generate_kmer_color_palette(top_outlier_kmers)
    kmer_plot <- create_kmer_plot(kmer_outlier_df, "kmer", outlier_color_palette)
    print(kmer_plot)
  }

  # Add Coverage and group summaries
  cat("\n\n")
  cat("### {-}") # This is a placeholder to stop the tabset

  cat("\n\n")
  cat("#### Count estimates")
  cat("\n\n")

  signif_par_summary_region_df %>%
    # Round to 2 decimals
    mutate_if(is.numeric, ~ round(., 2)) %>%
    # Select parameters, round and print
    select(
      "Locus ID" = locus_id,
      "Satellite" = satellite,
      "Haplotype" = haplotype,
      "Count estimate" = mean,
      "95% CI Lower" = mean_lower,
      "95% CI Upper" = mean_upper,
    ) %>%
    kbl(
      align = "c"
    ) %>%
    column_spec(1, italic = TRUE) %>%
    kable_styling("bordered", full_width = TRUE) %>%
    row_spec(
      0,
      background = "grey90"
    ) %>%
    collapse_rows(columns = c(1, 2), valign = "middle", target = 1) %>%
    cat()

  cat("\n\n")
  cat("#### Group summaries")
  cat("\n\n")

  clustering_summary_region_df %>%
    # Arrange by satellite and haplotype
    arrange(idx, haplotype) %>%
    # Select parameters, round and print
    select(
      "Satellite" = satellite,
      "Haplotype" = haplotype,
      "Coverage" = n,
      "Mean" = mean,
      "SD" = sd,
      "Median" = median,
      "IQR" = iqr,
    ) %>%
    mutate_if(is.numeric, ~ round(., 2)) %>%
    kbl(
      align = "c"
    ) %>%
    kable_styling("bordered", full_width = TRUE) %>%
    row_spec(
      0,
      background = HEADER_BACKGROUND
    ) %>%
    collapse_rows(columns = 1, valign = "middle") %>%
    cat()

  cat("\n\n")
  cat("### Additional read information {.tabset .tabset-pills}")
  cat("\n\n")

  cat("\n\n")
  cat("#### Passed (N=", n_haplotyped_reads, ")", sep = "")
  cat("\n\n")

  passed_reads_region_df <- reads_region_df %>%
    filter(haplotype != "outlier")

  if (nrow(passed_reads_region_df) == 0) {
    cat("There are no haplotyped reads for this locus.")
  } else {
    passed_reads_region_df %>%
      select(
        "Haplotype" = haplotype,
        "Read ID" = read_id,
        "Query Name" = query_name,
        "Alignment Type" = alignment_type,
        "Median STR Qual" = median_str_qual,
        "Kmer Count" = kmer_count_str
      ) %>%
      mutate_if(is.numeric, ~ round(., 3)) %>%
      kbl(
        align = "c"
      ) %>%
      kable_styling(c("bordered", "striped"), full_width = TRUE) %>%
      row_spec(
        0,
        background = HEADER_BACKGROUND
      ) %>%
      collapse_rows(columns = 1, valign = "middle") %>%
      cat()
  }

  cat("\n\n")
  cat("#### Outliers (N=", n_outlier_reads, ")", sep = "")
  cat("\n\n")

  if (no_outlier_reads) {
    cat("There are no outliers for this locus.")
  } else {
    reads_region_df %>%
      filter(haplotype == "outlier") %>%
      # Change "," to newline in outlier_reasons
      mutate(
        outlier_reasons = str_replace_all(outlier_reasons, ";", "\n")
      ) %>%
      select(
        "Read ID" = read_id,
        "Query Name" = query_name,
        "Alignment Type" = alignment_type,
        "Outlier Reasons" = outlier_reasons,
        "Median STR Qual" = median_str_qual,
        "Kmer Count" = kmer_count_str
      ) %>%
      mutate_if(is.numeric, ~ round(., 3)) %>%
      kbl(
        align = "c"
      ) %>%
      kable_styling(c("bordered", "striped"), full_width = TRUE) %>%
      row_spec(
        0,
        background = HEADER_BACKGROUND
      ) %>%
      cat()
  }

  cat("\n\n")
  cat("#### Filtered reads (N=", n_filtered_reads, ")", sep = "")
  cat("\n\n")

  if (nrow(filtered_reads_region_df) > 0) {
    filtered_reads_region_df %>%
      select(
        "Query Name" = query_name,
        "Error flags" = error_flags
      ) %>%
      kbl(
        align = "c"
      ) %>%
      kable_styling(c("bordered", "striped"), full_width = TRUE) %>%
      row_spec(
        0,
        background = HEADER_BACKGROUND
      ) %>%
      cat()
  } else {
    cat("There is no filtered reads for this locus. \n")
  }

  cat("\n\n")
  cat("### Consensus")
  cat("\n\n")

  consensus_region_df <- consensus_df %>%
    semi_join(selected_region, by = join_by(locus_id)) %>%
    wrangle_reads_data() %>%
    arrange(haplotype)

  cat("From each haplotype, a consensus sequence is generated. The consensus sequence is created as the 'median' of the observed sequences. The consensus sequence is shown in the table below, together an abacus plot showing the inividual $k$-mers.")

  # Create Abacus plots

  if (nrow(consensus_region_df) > 0) {
    consensus_kmer_df <- consensus_region_df %>%
      wrangle_reads_to_kmers()

    consensus_kmer_plot <- create_kmer_plot(consensus_kmer_df, "kmer", kmer_color_palette)
    print(consensus_kmer_plot)
  } else {
    cat("There is no consensus sequence for this locus.")
  }

  # Consensus table
  cat("\n\n")
  cat("#### Consensus sequences")
  cat("\n\n")

  if (nrow(consensus_region_df) > 0) {
    consensus_region_df %>%
      mutate(
        consensus_strings = add_colors_to_consensus_strings(consensus_strings, kmer_color_palette)
      ) %>%
      select(
        "Haplotype" = haplotype,
        "Read ID" = read_id,
        "Alignment Type" = alignment_subtype_txt,
        "Consensus" = consensus_strings
      ) %>%
      kbl(
        align = "c", escape = FALSE
      ) %>%
      kable_styling(c("bordered", "striped"), full_width = TRUE) %>%
      row_spec(
        0,
        background = HEADER_BACKGROUND
      ) %>%
      collapse_rows(columns = 1, valign = "middle") %>%
      cat()
  } else {
    cat("There is no consensus sequence for this locus. \n")
  }

  cat("\n\n")
  cat("### Histogram plots {.tabset .tabset-pills}")
  cat("\n\n")

  cat("A histogram showing the number of occurences of the individual $k$-mers. Is more than one $k$-mer is present in the pattern one histogram is shown for each $k$-mer. Reads that are removed as outliers are not shown in the plot.")

  if (no_haplotype_reads) {
    cat("There are no haplotyped reads for this locus.")
  } else {
    # Plot count


    # Get parameters
    em_het_par <- par_summary_region_df %>%
      filter(haplotype %in% c("h1", "h2")) %>%
      # Calculate sd
      mutate(
        sd = sqrt(unit_var * mean),
        satellite_idx = idx + 1
      )

    em_hom_par <- par_summary_region_df %>%
      filter(haplotype == "hom") %>%
      # Calculate sd
      mutate(
        sd = sqrt(unit_var * mean),
        satellite_idx = idx + 1
      )

    # Create density data
    density_df <-
      bind_rows(em_het_par, em_hom_par) %>%
      # Remove NA values
      filter(!is.na(mean)) %>%
      rowwise() %>%
      mutate(
        plot_sd = pmax(sd, 0.2),
        x = list(seq(mean - 5 * plot_sd, mean + 5 * plot_sd, length.out = 101)),
      ) %>%
      unnest(x) %>%
      group_by(haplotype) %>%
      mutate(
        d = dnorm(x, mean = mean, sd = plot_sd)
      )

    # Get length data
    length_df <- reads_region_df %>%
      mutate(
        count = str_split(kmer_count_str, "-")
      ) %>%
      unnest_longer(col = count, indices_to = "satellite_idx") %>%
      mutate(
        count = as.integer(count),
        alignment_type_txt = if_else(
          alignment_type == "spanning", "Spanning", "Flanking"
        )
      ) %>%
      select(
        haplotype, alignment_type_txt, satellite_idx, count
      )

    # Get satellites
    satellites <- str_split(selected_region$satellites_str, "-", simplify = TRUE)

    # Create plots for each satellite
    for (j in seq_along(satellites)) {
      # Get satellite
      satellite <- satellites[j]

      cat("\n\n")
      cat("#### Satellite", j, "{.tabset .tabset-pills}")
      cat("\n\n")

      # Filter data
      counts_index_df <- length_df %>%
        filter(
          satellite_idx == !!j
        ) %>%
        count(
          haplotype, alignment_type_txt, count
        )

      density_index_df <-
        density_df %>%
        filter(satellite_idx == !!j)

      cat("\n\n")
      cat("##### Grouped")
      cat("\n\n")

      length_plot_df <- counts_index_df %>%
        filter(haplotype != "outlier")
      if (nrow(length_plot_df) > 0) {
        length_plot <- generate_length_plot(length_plot_df, density_index_df, satellite, BIN_SZ)

        cat("\n\n")
        print(length_plot)
        cat("\n\n")
      } else {
        cat("\n\n")
        cat("There are no grouped reads for this locus.")
        cat("\n\n")
      }

      cat("\n\n")
      cat("##### Outliers")
      cat("\n\n")

      length_plot_df <- counts_index_df %>%
        filter(haplotype == "outlier")

      if (nrow(length_plot_df) > 0) {
        length_plot <- generate_length_plot(length_plot_df, data.frame(), satellite, BIN_SZ)
        cat("\n\n")
        print(length_plot)
        cat("\n\n")
      } else {
        cat("\n\n")
        cat("There are no outliers for this locus.")
        cat("\n\n")
      }
    }
  }

  cat("\n\n")
  cat("### Heterozygozity test summary")
  cat("\n\n")

  cat("This table shows the result from a statistical test for heterozygozity. The hypothesis ($H_A$) is that the sample is heterozygote i.e. that the data comes from two discrete normal distributions and the null hypothesis ($H_0$) is that the sample is homozygote i.e. that the data can be grouped into one gaussian cluster. The test performed is a log likelihood ratio test, where the statistic is assumed to be $\\chi^2$-distributed where the degrees of freedom is the difference in free paramteres under the two hypotheses. The value for the maximized log likelihood functions and the number of free parameteres under the two hypotheses, the test staistic, degrees of freedom, $p$-value and conclusion is shown in the following table.")

  test_summary_region_df %>%
    mutate(
      conclusion = if_else(
        is_significant,
        "Heterozygote",
        "Homozygote"
      )
    ) %>%
    select(
      "Log-likelihood Hetero" = log_lik_hetero,
      "Log-likelihood Hom" = log_lik_hom,
      "# Parameters Hetero" = n_par_hetero,
      "# Parameters Hom" = n_par_hom,
      "$\\chi^2$ Statistic" = statistic,
      "P-value" = p_value,
      "Significant" = is_significant,
      "Conclusion" = conclusion
    ) %>%
    kbl(
      align = "c"
    ) %>%
    kable_styling("bordered", full_width = TRUE) %>%
    row_spec(
      0,
      background = HEADER_BACKGROUND
    ) %>%
    cat()

  cat("\n\n")
  cat("#### Maximum likelihood estimates")
  cat("\n\n")

  cat("\n\n")
  cat("##### Mean")
  cat("\n\n")

  par_summary_region_df %>%
    mutate(
      model = if_else(haplotype == "hom", "Homozygote", "Heterozygote")
    ) %>%
    arrange(model, haplotype, idx) %>%
    select(
      "Model" = model,
      "Satellite" = satellite,
      "Haplotype" = haplotype,
      "Mean" = mean,
    ) %>%
    mutate_if(is.numeric, ~ round(., 2)) %>%
    kbl(align = "c") %>%
    kable_styling("bordered", full_width = TRUE) %>%
    collapse_rows(columns = 1:2, valign = "middle", target = 1) %>%
    row_spec(
      0,
      background = HEADER_BACKGROUND
    ) %>%
    cat()

  cat("\n\n")
  cat("##### Pi")
  cat("\n\n")

  par_summary_region_df %>%
    mutate(
      model = if_else(haplotype == "hom", "Homozygote", "Heterozygote")
    ) %>%
    arrange(model, haplotype, idx) %>%
    filter(
      haplotype %in% c("h1", "h2")
    ) %>%
    select(
      "Model" = model,
      "Haplotype" = haplotype,
      "Pi" = pi
    ) %>%
    distinct() %>%
    mutate_if(is.numeric, ~ round(., 2)) %>%
    kbl(align = "c") %>%
    kable_styling("bordered", full_width = TRUE) %>%
    collapse_rows(columns = 1, valign = "middle") %>%
    row_spec(
      0,
      background = HEADER_BACKGROUND
    ) %>%
    cat()


  cat("\n\n")
  cat("##### Variance")
  cat("\n\n")

  par_summary_region_df %>%
    mutate(
      model = if_else(haplotype == "hom", "Homozygote", "Heterozygote")
    ) %>%
    arrange(model, haplotype, idx) %>%
    select(
      "Model" = model,
      "Satellite" = satellite,
      "Unit variance" = unit_var
    ) %>%
    distinct() %>%
    mutate_if(is.numeric, ~ round(., 2)) %>%
    kbl(align = "c") %>%
    kable_styling("bordered", full_width = TRUE) %>%
    collapse_rows(columns = 1, valign = "middle") %>%
    row_spec(
      0,
      background = HEADER_BACKGROUND
    ) %>%
    cat()
}
```
