---
title: "Short Tandem Repeat"
output: 
  html_document:
    toc: FALSE 
    code_folding: hide

params:
  sample_id: ""
  input_bam: ""
  str_catalouge: ""
  reads_csv: ""
  consensus_csv: ""
  filtered_reads_csv: ""
  clustering_summary_csv: ""
  em_summary_csv: ""
  par_summary_csv: ""
---

```{r Setup, include=FALSE}
# Setup
knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  results = FALSE,
  out.width = "100%",
  fig.align = "center"
)

options(knitr.table.format = "html")

# Load libraries
library(tidyverse)
library(RColorBrewer)
library(knitr)
library(kableExtra)
```

```{r Unpack input parameters}
# Sample information
sample_id <- params$sample_id
input_bam <- params$input_bam
str_catalouge <- params$str_catalouge

# Unpack input variables
reads_csv <- params$reads_csv
filtered_reads_csv <- params$filtered_reads_csv
consensus_csv <- params$consensus_csv
clustering_summary_csv <- params$clustering_summary_csv
em_summary_csv <- params$em_summary_csv
par_summary_csv <- params$par_summary_csv

print(params)

# TODO: Make variable names more descriptive
# Load input files
reads_df <- read_csv(reads_csv)
filtered_reads_df <- read_csv(filtered_reads_csv)
consensus_df <- read_csv(consensus_csv)

clustering_summary_df <- read_csv(clustering_summary_csv)
em_summary_df <- read_csv(em_summary_csv)
par_summary_df <- read_csv(par_summary_csv)
```

```{r Test exmaple data, eval=FALSE}
# Load data
reads_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.read_info.csv"
reads_df <- read_csv(reads_csv)

filtered_reads_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.filtered_reads_info.csv"
filtered_reads_df <- read_csv(filtered_reads_csv)

clustering_summary_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.haplotype_info.csv"
clustering_summary_df <- read_csv(clustering_summary_csv)

em_summary_csv <- "/faststorage/project/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/test.summary.csv"
em_summary_df <- read_csv(em_summary_csv)
```

```{r Test exmaple data - local, eval=FALSE}
# Load data
reads_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/reads.csv"
reads_df <- read_csv(reads_csv)

filtered_reads_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/filtered_reads.csv"
filtered_reads_df <- read_csv(filtered_reads_csv)

consensus_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/consensus.csv"
consensus_df <- read_csv(consensus_csv)

clustering_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/haplotypes.csv"
clustering_summary_df <- read_csv(clustering_summary_csv)

em_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/summary.csv"
em_summary_df <- read_csv(em_summary_csv)

par_summary_csv <- "~/GenomeDK/MomaNanoporeDevelopment/BACKUP/devel/simond/abacus/testing/abacus_output/par_summary.csv"
par_summary_df <- read_csv(par_summary_csv)
```


```{r Get destinct region}
# Get df of region information
regions <- reads_df %>%
  distinct(locus_id, locus_chrom, locus_start, locus_end, satellites_str, structure)
```

```{r Sample information table, results = TRUE}
# Create table
data.frame(
  sample_id = sample_id,
  date = format(Sys.time(), "%Y/%m/%d %H:%M %Z"),
  bam = basename(input_bam),
  json = basename(str_catalouge)
) %>%
  select(
    "Sample ID" = sample_id,
    "Date" = date,
    "BAM file" = bam,
    "STR catalouge" = json
  ) %>%
  # Transpose
  t() %>%
  kbl(
    align = "lc"
  ) %>%
  kable_styling("bordered", full_width = T) %>%
  column_spec(1, bold = T)
```

```{r Report, results = "asis"}
cat("\n\n")
cat("# Choose a region {.tabset .tabset-pills}")
cat("\n\n")

for (i in seq_len(nrow(regions))) {
  selected_region <- regions[i, ]

  # Filter data
  reads_region_df <- reads_df %>%
    semi_join(selected_region, by = join_by(locus_id)) %>%
    # Get number of kmers
    mutate(
      # Fix missing
      obs_kmer_string = if_else(is.na(obs_kmer_string), "", obs_kmer_string),
      n_kmer = str_count(obs_kmer_string, "-") + 1
    ) %>%
    # Get max per em_haplotype
    group_by(em_haplotype) %>%
    mutate(max_kmer_haplotype = max(n_kmer)) %>%
    ungroup() %>%
    # Correct flanking reads
    mutate(
      start = case_when(
        alignment_type == "left_flanking" ~ -Inf,
        alignment_type == "right_flanking" ~ max_kmer_haplotype - n_kmer + 1,
        alignment_type == "spanning" ~ -Inf
      ),
      end = case_when(
        alignment_type == "left_flanking" ~ n_kmer,
        alignment_type == "right_flanking" ~ Inf,
        alignment_type == "spanning" ~ Inf
      ),
      # Convert alignment_type to factor
      alignment_type = factor(alignment_type, levels = c("left_flanking", "spanning", "right_flanking"))
    ) %>%
    ungroup() %>%
    # Arrange reads by em_haplotype, alignment_type, mapping_start, length and sequence
    arrange(em_haplotype, alignment_type, start, n_kmer, obs_kmer_string) %>%
    mutate(read_id = row_number())

  if (nrow(filtered_reads_df) > 0) {
    filtered_reads_region_df <- filtered_reads_df %>%
      semi_join(selected_region, by = join_by(locus_id))
  } else {
    filtered_reads_region_df <- data.frame()
  }

  em_summary_region_df <- em_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))

  par_summary_region_df <- par_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))

  res_region_df <- clustering_summary_df %>%
    semi_join(selected_region, by = join_by(locus_id))


  # Create heading - Note this is a tab
  cat("\n\n")
  cat("##", selected_region$locus_id)
  cat("\n\n")

  selected_region %>%
    mutate(
      goto_link = paste0("http://localhost:60151/goto?locus=", locus_chrom, ":", locus_start, "-", locus_end),
      igv_link = paste0("[Go to locus](", goto_link, ")"),
      structure = str_replace_all(structure, fixed("*"), fixed("\\*"))
    ) %>%
    select(
      "Locus ID" = locus_id,
      "Chrom" = locus_chrom,
      "Start" = locus_start,
      "End" = locus_end,
      "satellites" = satellites_str,
      "Structure" = structure,
      "IGV link" = igv_link
    ) %>%
    kbl(
      align = "c"
    ) %>%
    kable_styling(c("bordered", "striped"), full_width = T) %>%
    cat()

  cat("\n\n")
  cat("### Coverage summary")
  cat("\n\n")

  data.frame(
    total_coverage = nrow(reads_region_df) + nrow(filtered_reads_region_df),
    filtered_reads = nrow(filtered_reads_region_df),
    analyzed_reads = nrow(reads_region_df),
    outliers = nrow(reads_region_df %>% filter(em_haplotype == "outlier")),
    haplotyped_reads = nrow(reads_region_df %>% filter(em_haplotype != "outlier"))
  ) %>%
    select(
      "Total coverage" = total_coverage,
      "Filtered" = filtered_reads,
      "Analyzed" = analyzed_reads,
      "Outliers" = outliers,
      "Used for haplotyping" = haplotyped_reads,
    ) %>%
    kbl(
      align = "c"
    ) %>%
    kable_styling(c("bordered", "striped"), full_width = T) %>%
    cat()

  cat("\n\n")
  cat("### Heterozygozity test summary")
  cat("\n\n")

  cat("This table shows the result from a statistical test for heterozygozity. The hypothesis ($H_A$) is that the sample is heterozygote i.e. that the data comes from two discrete normal distributions and the null hypothesis ($H_0$) is that the sample is homozygote i.e. that the data can be grouped into one gaussian cluster. The test performed is a log likelihood ratio test, where the statistic is assumed to be $\\chi^2$-distributed where the degrees of freedom is the difference in free paramteres under the two hypotheses. The value for the maximized log likelihood functions and the number of free parameteres under the two hypotheses, the test staistic, degrees of freedom, $p$-value and conclusion is shown in the following table.")

  em_summary_region_df %>%
    mutate(
      conclusion = if_else(
        is_significant,
        "Heterozygote",
        "Homozygote"
      )
    ) %>%
    select(
      "Log-likelihood Hetero" = log_lik_hetero,
      "Log-likelihood Hom" = log_lik_hom,
      "# Parameters Hetero" = n_par_hetero,
      "# Parameters Hom" = n_par_hom,
      "$\\chi^2$ Statistic" = statistic,
      "P-value" = p_value,
      "Significant" = is_significant,
      "Conclusion" = conclusion
    ) %>%
    kbl(
      align = "c"
    ) %>%
    kable_styling(c("bordered", "striped"), full_width = T) %>%
    cat()

  cat("\n\n")
  cat("#### Maximum likelihood estimates")
  cat("\n\n")

  par_summary_region_df %>%
    # Arrange by satellite and haplotype
    arrange(em_haplotype, idx) %>%
    # Select parameters, round and print
    select(
      "Haplotype" = em_haplotype,
      "Pi" = pi,
      "Satellite" = satellite,
      "Mean" = mean,
      "Unit variance" = unit_var,
    ) %>%
    mutate_if(is.numeric, ~ round(., 2)) %>%
    kbl(
      align = "c"
    ) %>%
    column_spec(1, italic = T) %>%
    column_spec(3, italic = T) %>%
    kable_styling(c("bordered", "striped"), full_width = T) %>%
    collapse_rows(columns = c(1, 2), valign = "middle") %>%
    cat()

  cat("\n\n")
  cat("#### Group summaries")
  cat("\n\n")

  res_region_df %>%
    # Arrange by satellite and haplotype
    arrange(em_haplotype, satellite) %>%
    # Select parameters, round and print
    select(
      "Haplotype" = em_haplotype,
      "Count" = n,
      "Satellite" = satellite,
      "Mean" = mean,
      "SD" = sd,
      "Median" = median,
      "IQR" = iqr,
    ) %>%
    mutate_if(is.numeric, ~ round(., 2)) %>%
    kbl(
      align = "c"
    ) %>%
    column_spec(1, italic = T) %>%
    column_spec(3, italic = T) %>%
    kable_styling(c("bordered", "striped"), full_width = T) %>%
    collapse_rows(columns = c(1, 2), valign = "middle") %>%
    cat()

  # Abacus plot
  cat("\n\n")
  cat("### Abacus plot {.tabset .tabset-pills}")
  cat("\n\n")

  cat("\n\n")
  cat("A visual representation illustrating the distribution of individual $k$-mers as colored *beads* along a linear sequence, with each horizontal sequence representing individual reads. The reads are grouped into haplotypes called by an unsupervised clustering algorithm (EM). Additionally, a separate group is designated for outlier reads. A read can be classified as an outlier either based on an unusual $k$-mer count or nucleotide sequence. There are 3 versions of the plot where the beads are colored by the observed $k$-mer, the reference $k$-mer, and the methylation probability of the $k$-mer. Spanning and flanking reads are shown as black and grey lines, respectively.")
  cat("\n\n")

  # Wrangle STR data for abacus plot
  mod_ascii_to_prob <- function(mod_ascii) {
    # Check if mod_ascii is empty
    if (nchar(mod_ascii) == 0) {
      return(0)
    }
    # Convert to numeric
    mod_prob <- mod_ascii %>%
      charToRaw() %>%
      as.numeric()
    # Correct for ASCII values
    mod_prob <- (mod_prob - 33) * 10
    # Remove values outside 0-100
    mod_prob <- mod_prob[mod_prob >= 0 & mod_prob <= 100]
    # Return NA if no values
    if (length(mod_prob) == 0) {
      return(NA)
    }
    # Set mod_prob to max value
    mod_prob <- max(mod_prob)
    return(mod_prob)
  }

  kmer_df <- reads_region_df %>%
    # Get one line per kmer
    mutate(
      kmer = str_split(obs_kmer_string, "-"),
      ref_kmer = str_split(ref_kmer_string, "-"),
      mod_kmer = str_split(mod_5mc_kmer_string, "-")
    ) %>%
    unnest_longer(
      col = c(kmer, ref_kmer, mod_kmer),
      values_to = "{col}",
      indices_to = "{col}_idx"
    ) %>%
    # Fix NA values
    mutate(
      kmer = if_else(is.na(kmer), "", kmer),
      ref_kmer = if_else(is.na(ref_kmer), "", ref_kmer),
      mod_kmer = if_else(is.na(mod_kmer), "", mod_kmer)
    ) %>%
    # Translated mod kmer to probability from ASCII value (33-43 -> 0,10,20,...,100)
    rowwise() %>%
    mutate(
      mod_prob = mod_ascii_to_prob(mod_kmer)
    )

  # Make color palette for kmers
  generate_kmer_color_palette <- function(kmers) {
    # Set max number of colors
    max_cols <- 12

    # Trim kmers if too many
    if (length(kmers) > max_cols) {
      kmers <- kmers[1:max_cols]
    }

    # Get color palette
    full_paired_pallete <- brewer.pal(max_cols, "Paired")
    reordered_pallete <- c(full_paired_pallete[(1:6) * 2], full_paired_pallete[(1:6) * 2 - 1])

    kmer_colors <- reordered_pallete[seq_along(kmers)]
    names(kmer_colors) <- kmers

    return(kmer_colors)
  }

  # Function to create kmer plots
  create_kmer_plot <- function(kmer_df, color_by, kmer_color_palette = NULL) {
    # Fix data
    kmer_plot_df <- kmer_df %>%
      mutate(
        # Correct stqrt of right flanking reads
        cor_kmer_idx = if_else(
          alignment_type == "right_flanking",
          kmer_idx + start - 1,
          kmer_idx
        ),
        alignment_type_txt = if_else(
          alignment_type == "spanning", "Spanning", "Flanking"
        )
      ) %>%
      ungroup()

    # Start plotting
    plot <- ggplot()

    # Add horizontal lines for each read
    hz_lines_df <- kmer_plot_df %>%
      distinct(read_id, alignment_type_txt, start, end, em_haplotype)

    plot <- plot +
      geom_segment(
        aes(
          x = start, xend = end,
          y = read_id, yend = read_id,
          alpha = alignment_type_txt
        ),
        lwd = 1,
        data = hz_lines_df
      ) +
      scale_alpha_manual(values = c("Spanning" = 1, "Flanking" = 0.4)) +
      labs(alpha = "Alignment type")

    # Get "dimensionality" of plot
    n_reads <- nrow(hz_lines_df)
    n_kmers <- max(kmer_plot_df$cor_kmer_idx)

    # Add points for each kmer
    if (color_by == "kmer" || color_by == "ref_kmer") {
      # If ref_kmer is selected, use ref_kmer instead of kmer
      if (color_by == "ref_kmer") {
        kmer_plot_df <- kmer_plot_df %>%
          mutate(kmer = ref_kmer)
      }

      # Add "Other" (grey) to color palette in end
      kmer_color_palette <- c(kmer_color_palette, c("Other" = "grey"))

      kmer_plot_df <- kmer_plot_df %>%
        mutate(
          # Change non-top kmers to "Other"
          kmer = if_else(kmer %in% names(kmer_color_palette), kmer, "Other"),
          # Convert to factor
          kmer = factor(kmer, levels = names(kmer_color_palette))
        )

      # Add points/tiles and color by kmer
      if (n_reads < 50 && n_kmers < 60) {
        plot <- plot +
          geom_point(
            aes(
              x = cor_kmer_idx, y = read_id,
              col = kmer
            ),
            size = 5,
            data = kmer_plot_df
          ) +
          # Add color palette
          scale_color_manual(values = kmer_color_palette) +
          labs(color = "K-mer\n(ord. by freq.)")
      } else {
        plot <- plot +
          geom_tile(
            aes(
              x = cor_kmer_idx, y = read_id,
              fill = kmer
            ),
            data = kmer_plot_df
          ) +
          # Add color palette
          scale_fill_manual(values = kmer_color_palette) +
          labs(fill = "K-mer\n(ord. by freq.)")
      }
    } else if (color_by == "mod") {
      # Add points/tiles and color by mod probability
      if (n_reads < 50 && n_kmers < 60) {
        plot <- plot +
          geom_point(
            aes(x = cor_kmer_idx, y = read_id, col = mod_prob),
            size = 5,
            data = kmer_plot_df
          ) +
          # Add color gradient with low (0) = "blue", high (100) = "red", NA = "grey" (similar to IGV)
          scale_color_gradient(
            low = "blue",
            high = "red",
            na.value = "grey",
            limits = c(0, 100)
          ) +
          labs(color = "5mC (prob.)")
      } else {
        plot <- plot +
          geom_tile(
            aes(x = cor_kmer_idx, y = read_id, fill = mod_prob),
            data = kmer_plot_df
          ) +
          # Add color gradient with low (0) = "blue", high (100) = "red", NA = "grey" (similar to IGV)
          scale_fill_gradient(
            low = "blue",
            high = "red",
            na.value = "grey",
            limits = c(0, 100)
          ) +
          labs(fill = "5mC (prob.)")
      }
    } else {
      # Error
      stop("Invalid color_by argument.")
    }

    # Add details to plot
    min_kmer_idx <- min(kmer_plot_df$cor_kmer_idx)
    plot <- plot +
      facet_grid(
        rows = vars(em_haplotype),
        scales = "free_y",
        space = "free_y"
      ) +
      labs(
        title = "K-mer overview",
        x = "K-mer index",
        y = "Read ID",
      ) +
      # Breaks for y: Every integer, pretty, start at 1
      scale_y_reverse(breaks = function(x) unique(round(sort(pmax(min_kmer_idx, pretty(x))))))

    return(plot)
  }

  # Split data
  kmer_haplotyped_df <- kmer_df %>%
    filter(em_haplotype != "outlier")

  kmer_outlier_df <- kmer_df %>%
    filter(em_haplotype == "outlier")

  no_haplotype_reads <- nrow(kmer_haplotyped_df) == 0
  no_outlier_reads <- nrow(kmer_outlier_df) == 0

  # TODO
  # Get color palette
  # Get kmers that should be prioritized for color
  # Ref kmers
  ref_kmers <- kmer_haplotyped_df %>%
    filter(!is.na(ref_kmer)) %>%
    count(ref_kmer) %>%
    arrange(desc(n)) %>%
    pull(ref_kmer)

  # Most observed kmers
  top_obs_kmers <-
    kmer_haplotyped_df %>%
    # Remove outlier reads, NA and empty kmers
    filter(em_haplotype != "outlier", !is.na(kmer), kmer != "") %>%
    count(kmer) %>%
    # Remove unique kmers (assumed to be errors)
    filter(n > 1) %>%
    # Get top kmers
    arrange(desc(n)) %>%
    pull(kmer) %>%
    # Remove kmers that are in ref_kmers
    setdiff(ref_kmers)

  # Select kmers for color palette
  n_colors <- 12
  top_kmers <- c(ref_kmers, top_obs_kmers) %>%
    unique() %>%
    head(n_colors)

  kmer_color_palette <- generate_kmer_color_palette(top_kmers)

  cat("\n\n")
  cat("#### Sequence")
  cat("\n\n")

  if (no_haplotype_reads) {
    cat("There are no haplotyped reads for this locus.")
  } else {
    kmer_plot <- create_kmer_plot(kmer_haplotyped_df, "kmer", kmer_color_palette)
    print(kmer_plot)
  }


  cat("\n\n")
  cat("#### Reference")
  cat("\n\n")

  if (no_haplotype_reads) {
    cat("There are no haplotyped reads for this locus.")
  } else {
    kmer_plot <- create_kmer_plot(kmer_haplotyped_df, "ref_kmer", kmer_color_palette)
    print(kmer_plot)
  }

  cat("\n\n")
  cat("#### Methylation (5mC)")
  cat("\n\n")

  if (no_haplotype_reads) {
    cat("There are no haplotyped reads for this locus.")
  } else if (all(is.na(kmer_haplotyped_df$mod_prob))) {
    cat("There are no methylation information for this locus.")
  } else {
    kmer_plot <- create_kmer_plot(kmer_haplotyped_df, "mod", 11)
    print(kmer_plot)
  }

  cat("\n\n")
  cat("#### Outliers")
  cat("\n\n")

  if (no_outlier_reads) {
    cat("There are no outliers for this locus.")
  } else {
    # Outlier color pallette
    top_outlier_kmers <- kmer_outlier_df %>%
      filter(!is.na(kmer)) %>%
      count(kmer) %>%
      arrange(desc(n)) %>%
      pull(kmer) %>%
      head(n_colors)

    outlier_color_palette <- generate_kmer_color_palette(top_outlier_kmers)
    kmer_plot <- create_kmer_plot(kmer_outlier_df, "kmer", outlier_color_palette)
    print(kmer_plot)
  }

  cat("\n\n")
  cat("### Additional read information {.tabset .tabset-pills}")
  cat("\n\n")

  cat("\n\n")
  cat("#### Passed")
  cat("\n\n")

  cat("A table with additional information for the reads visualized above.")

  reads_region_df %>%
    filter(em_haplotype != "outlier") %>%
    select(
      "Haplotype" = em_haplotype,
      "Read ID" = read_id,
      "Query Name" = query_name,
      "Alignment Type" = alignment_type,
      "Median STR Qual" = median_str_qual,
      "Kmer Count" = kmer_count_str
    ) %>%
    mutate_if(is.numeric, ~ round(., 3)) %>%
    kbl(
      align = "c"
    ) %>%
    column_spec(1, italic = T) %>%
    kable_styling(c("bordered", "striped"), full_width = T) %>%
    collapse_rows(columns = 1, valign = "middle") %>%
    cat()

  cat("\n\n")
  cat("#### Outliers")
  cat("\n\n")

  reads_region_df %>%
    filter(em_haplotype == "outlier") %>%
    # Change "," to newline in outlier_reasons
    mutate(
      outlier_reasons = str_replace_all(outlier_reasons, ";", "\n")
    ) %>%
    select(
      "Read ID" = read_id,
      "Query Name" = query_name,
      "Alignment Type" = alignment_type,
      "Outlier Reasons" = outlier_reasons,
      "Median STR Qual" = median_str_qual,
      "Kmer Count" = kmer_count_str
    ) %>%
    mutate_if(is.numeric, ~ round(., 3)) %>%
    kbl(
      align = "c"
    ) %>%
    kable_styling(c("bordered", "striped"), full_width = T) %>%
    cat()

  cat("\n\n")
  cat("#### Filtered")
  cat("\n\n")

  cat("A table with information about the filtered reads and the reason why they were removed.\n")

  if (nrow(filtered_reads_region_df) > 0) {
    filtered_reads_region_df %>%
      select(
        "Query Name" = query_name,
        "Error flags" = error_flags
      ) %>%
      kbl(
        align = "c"
      ) %>%
      kable_styling(c("bordered", "striped"), full_width = T) %>%
      cat()
  } else {
    cat("There is no filtered reads for this locus. \n")
  }

  cat("\n\n")
  cat("### Consensus {.tabset .tabset-pills}")
  cat("\n\n")

  consensus_region_df <- consensus_df %>%
    semi_join(selected_region, by = join_by(locus_id)) %>%
    arrange(em_haplotype)

  cat("From each haplotype, a consensus sequence is generated. The consensus sequence is created as the 'median' of the observed sequences. The consensus sequence is shown in the table below, together an abacus plot showing the inividual $k$-mers.")

  if (nrow(consensus_region_df) > 0) {
    consensus_region_df %>%
      select(
        "Haplotype" = em_haplotype,
        "Consensus" = consensus_strings
      ) %>%
      kbl(
        align = "c"
      ) %>%
      kable_styling(c("bordered", "striped"), full_width = T) %>%
      column_spec(1, italic = T) %>%
      cat()
  } else {
    cat("There is no consensus sequence for this locus. \n")
  }

  # Create Abacus plots

  consensus_kmer_df <- consensus_region_df %>%
    # Added to be compatible with create_kmer_plot
    mutate(
      start = -Inf,
      end = Inf,
      read_id = row_number()
    ) %>%
    # Get one line per kmer
    mutate(
      kmer = str_split(obs_kmer_string, "-"),
      ref_kmer = str_split(ref_kmer_string, "-"),
      mod_kmer = str_split(mod_5mc_kmer_string, "-")
    ) %>%
    unnest_longer(
      col = c(kmer, ref_kmer, mod_kmer),
      values_to = "{col}",
      indices_to = "{col}_idx"
    ) %>%
    # Fix NA values
    mutate(
      kmer = if_else(is.na(kmer), "", kmer),
      ref_kmer = if_else(is.na(ref_kmer), "", ref_kmer),
      mod_kmer = if_else(is.na(mod_kmer), "", mod_kmer)
    ) %>%
    # Translated mod kmer to probability from ASCII value (33-43 -> 0,10,20,...,100)
    rowwise() %>%
    mutate(
      mod_prob = mod_ascii_to_prob(mod_kmer),
    )

  cat("\n\n")
  cat("#### Sequence")
  cat("\n\n")

  if (nrow(consensus_region_df) > 0) {
    consensus_kmer_plot <- create_kmer_plot(consensus_kmer_df, "kmer", kmer_color_palette)
    print(consensus_kmer_plot)
  } else {
    cat("There is no consensus sequence for this locus.")
  }

  generate_length_plot <- function(counts_df, density_df, satellite, BIN_SZ) {
    # Bin count to min max 100 bins
    NBINS <- min(100, max(counts_df$count))
    BIN_SZ <- ceiling(max(counts_df$count) / NBINS)

    counts_df <- counts_df %>%
      mutate(
        count = count %/% BIN_SZ * BIN_SZ
      ) %>%
      group_by(alignment_type_txt, count) %>%
      summarise(n = sum(n), .groups = "drop")

    length_plot <- ggplot(counts_df) +
      # Add histogram of STR length
      geom_col(
        aes(
          x = count,
          y = n,
          fill = alignment_type_txt
        ),
        width = BIN_SZ
      ) +
      scale_x_continuous(breaks = function(x) unique(round(sort(pretty(x))))) +
      scale_y_continuous(breaks = function(x) unique(round(sort(pretty(x))))) +
      scale_fill_manual(values = c("Spanning" = "grey30", "Flanking" = "grey70"))

    # Add distributions
    if (nrow(density_df) > 0) {
      # Fix colors
      em_haplotype_colors <- brewer.pal(3, "Set1")
      names(em_haplotype_colors) <- c("hom", "h1", "h2")
      # Fix line type
      em_haplotype_linetype <- c("hom" = "dashed", "h1" = "solid", "h2" = "solid")

      # Scale density to max count
      if (nrow(density_df) > 0) {
        density_df <- density_df %>%
          group_by(em_haplotype) %>%
          mutate(
            d = d / max(d) * max(counts_df$n)
          ) %>%
          ungroup()
      }

      length_plot <- length_plot +
        geom_line(
          data = density_df,
          aes(x = x, y = d, col = em_haplotype, linetype = em_haplotype),
          linewidth = 0.5
        ) +
        scale_color_manual(values = em_haplotype_colors) +
        scale_linetype_manual(values = em_haplotype_linetype, guide = "none")
    }

    # Add title, labels and legend
    length_plot <- length_plot +
      # Add title and labels
      labs(
        title = paste("Satellite:", satellite),
        x = "STR length",
        y = "Count"
      ) +
      # Name the legend
      guides(
        fill = guide_legend(title = "Alignment type"),
        col = guide_legend(title = "Haplotype")
      )

    return(length_plot)
  }

  cat("\n\n")
  cat("### Histogram plots {.tabset .tabset-pills}")
  cat("\n\n")

  cat("A histogram showing the number of occurences of the individual $k$-mers. Is more than one $k$-mer is present in the pattern one histogram is shown for each $k$-mer. Reads that are removed as outliers are not shown in the plot.")

  if (no_haplotype_reads) {
    cat("There are no haplotyped reads for this locus.")
  } else {
    # Plot count


    # Get parameters
    em_het_par <- par_summary_region_df %>%
      filter(em_haplotype %in% c("h1", "h2")) %>%
      # Calculate sd
      mutate(
        sd = sqrt(unit_var * mean),
        satellite_idx = idx + 1
      )

    em_hom_par <- par_summary_region_df %>%
      filter(em_haplotype == "hom") %>%
      # Calculate sd
      mutate(
        sd = sqrt(unit_var * mean),
        satellite_idx = idx + 1
      )

    # Create density data
    density_df <-
      bind_rows(em_het_par, em_hom_par) %>%
      rowwise() %>%
      mutate(
        plot_sd = pmax(sd, 0.2),
        x = list(seq(mean - 5 * plot_sd, mean + 5 * plot_sd, length.out = 101)),
      ) %>%
      unnest(x) %>%
      group_by(em_haplotype) %>%
      mutate(
        d = dnorm(x, mean = mean, sd = plot_sd)
      )

    # Get length data
    length_df <- reads_region_df %>%
      mutate(
        count = str_split(kmer_count_str, "-")
      ) %>%
      unnest_longer(col = count, indices_to = "satellite_idx") %>%
      mutate(
        count = as.integer(count),
        alignment_type_txt = if_else(
          alignment_type == "spanning", "Spanning", "Flanking"
        )
      ) %>%
      select(
        em_haplotype, alignment_type_txt, satellite_idx, count
      )

    # Get satellites
    satellites <- str_split(selected_region$satellites_str, "-", simplify = TRUE)

    # Create plots for each satellite
    for (j in seq_along(satellites)) {
      # Get satellite
      satellite <- satellites[j]

      cat("\n\n")
      cat("#### Satellite", j, "{.tabset .tabset-pills}")
      cat("\n\n")

      # Filter data
      counts_index_df <- length_df %>%
        filter(
          satellite_idx == !!j
        ) %>%
        count(
          em_haplotype, alignment_type_txt, count
        )

      density_index_df <-
        density_df %>%
        filter(satellite_idx == !!j)

      cat("\n\n")
      cat("##### Grouped")
      cat("\n\n")

      length_plot_df <- counts_index_df %>%
        filter(em_haplotype != "outlier")
      if (nrow(length_plot_df) > 0) {
        length_plot <- generate_length_plot(length_plot_df, density_index_df, satellite, BIN_SZ)

        cat("\n\n")
        print(length_plot)
        cat("\n\n")
      } else {
        cat("\n\n")
        cat("There are no grouped reads for this locus.")
        cat("\n\n")
      }

      cat("\n\n")
      cat("##### Outliers")
      cat("\n\n")

      length_plot_df <- counts_index_df %>%
        filter(em_haplotype == "outlier")

      if (nrow(length_plot_df) > 0) {
        length_plot <- generate_length_plot(length_plot_df, data.frame(), satellite, BIN_SZ)
        cat("\n\n")
        print(length_plot)
        cat("\n\n")
      } else {
        cat("\n\n")
        cat("There are no outliers for this locus.")
        cat("\n\n")
      }
    }
  }
}
```
